src/BPSuckerDelegate.sol

```diff
@@ -0,0 +1,177 @@
+    IJBRulesets public immutable RULESETS;
+
+    /// @notice The contract that exposes price feeds.
+    IJBPrices public immutable PRICES;
+
+    constructor(
+        IJBPrices _prices,
+        IJBRulesets _rulesets,
+        OPMessenger _messenger,
+        IJBDirectory _directory,
+        IJBTokens _tokens,
+        IJBPermissions _permissions,
+        address _peer
+    ) BPOptimismSucker(_messenger, _directory, _tokens, _permissions, _peer) {
+        PRICES = _prices;
+        RULESETS = _rulesets;
+    }
+
+    /// @notice The data calculated before a payment is recorded in the terminal store. This data is provided to the
+    /// terminal's `pay(...)` transaction.
+    /// @param context The context passed to this data hook by the `pay(...)` function as a `JBBeforePayRecordedContext`
+    /// struct.
+    /// @return weight The new `weight` to use, overriding the ruleset's `weight`.
+    /// @return hookSpecifications The amount and data to send to pay hooks instead of adding to the terminal's balance.
+    function beforePayRecordedWith(JBBeforePayRecordedContext calldata context)
+        external
+        view
+        returns (uint256 weight, JBPayHookSpecification[] memory hookSpecifications)
+    {
+        uint256 _remoteProjectId = acceptFromRemote[context.projectId];
+        if (
+            context.amount.token != JBConstants.NATIVE_TOKEN
+            // We only support the native asset (for now).
+            // The remote project was not configured.
+            || _remoteProjectId == 0
+            // Check if the terminal supports the redeem terminal interface.
+            || !ERC165Checker.supportsInterface(address(context.terminal), type(IJBRedeemTerminal).interfaceId)
+        ) {
+            // In these cases we don't do anything.
+            return (context.weight, new JBPayHookSpecification[](0));
+        }
+
+        // Get the projects ruleset.
+        JBRuleset memory _ruleset = RULESETS.getRulesetOf(context.projectId, context.rulesetId);
+
+        // Calculate the amount of tokens that would be minted for this payment.
+        uint256 _weightRatio = context.amount.currency == _ruleset.baseCurrency()
+            ? 10 ** context.amount.decimals
+            : PRICES.pricePerUnitOf({
+                projectId: context.projectId,
+                pricingCurrency: context.amount.currency,
+                unitCurrency: _ruleset.baseCurrency(),
+                decimals: context.amount.decimals
+            });
+
+        uint256 _tokenCount = mulDiv(context.amount.value, context.weight, _weightRatio);
+
+        // We return zero weight, so that we can do the mint on the remote chain.
+        hookSpecifications = new JBPayHookSpecification[](1);
+        hookSpecifications[0] = JBPayHookSpecification({
+            hook: IJBPayHook(address(this)),
+            amount: 0,
+            metadata: abi.encode(
+                EncodedMetadata({
```

Comment:
wondering if we have to send metadata? the `context` on the after fn has access to terminal, all ingredients to make tokenCount, beneficiary, and we can do the remote projectId check there?

Reply:
that way we can keep `before` skinny, and do basically all the work in `after`